sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -86,6 +86,25 @@ trait DateTimeFormatterHelper {
         throw e
       }
   }
+
+  // When the new DateTimeFormatter failed to initialize because of invalid datetime pattern, it
+  // will throw IllegalArgumentException. If the pattern can be recognized by the legacy formatter
+  // it will raise SparkUpgradeException to tell users to restore the previous behavior via LEGACY
+  // policy or follow our guide to correct their pattern.
+  protected def checkLegacyFormatter[T1, T2](
```

Comment:
what does the type parameter give us? The return type doesn't depend on T1 or T2